# Multi-Platform Cosmetics Market Analysis | Nykaa Data Cleaning

Author: Neethu Raj

## Project Overview

This project analyzes the Indian online cosmetics market using product data collected from multiple e-commerce platforms. The current notebook focuses on the **Nykaa dataset** and documents the complete data-cleaning process applied to the scraped product listings.

The dataset is processed through a structured workflow that includes:

* Initial data exploration and quality assessment
* Missing value analysis and treatment
* Duplicate record and duplicate product investigation
* Extraction and construction of product identifiers
* Product-level deduplication
* Logical validation of prices, discounts, ratings, and essential fields
* Data type conversion and feature preparation
* Final dataset validation and export

A key objective of this cleaning process is to transform the raw scraped listings into a reliable **product-level dataset** suitable for downstream analysis and comparison with data collected from other e-commerce platforms.

The cleaned dataset produced in this notebook will serve as the Nykaa component of the broader multi-platform cosmetics market analysis project.


| Phase                                                |    Status   |
| ---------------------------------------------------- | :---------: |
| 1. Web Scraping – Nykaa                              | ✅ Completed |
| 2. Data Cleaning – Nykaa                             | 🔄 In Progress |
| 3. Exploratory Data Analysis  - Nykaa                |  ⏳ Planned  |
| 4. SQL Business Analysis - Nykaa                     |  ⏳ Planned  |
| 5. Power BI Dashboard - Nykaa                        |  ⏳ Planned  |
| 6. Web Scraping – Myntra                             |  ⏳ Planned  |
| 7. Data Cleaning – Myntraa                           |  ⏳ Planned  |
| 8. Cross-Platform Price & Discount Comparison        |  ⏳ Planned  |
| 9. Cross-Platform Power BI Dashboard                 |  ⏳ Planned  |



## Dataset Information

* **Source:** Self-scraped from Nykaa using Python.
* **Total records before cleaning:** 27,364
* **Number of features before cleaning:** 12
* **Platform:** Nykaa

The raw dataset contains the following features:

* **product_name** – Name of the product
* **url** – Product listing URL
* **category** – Main product category
* **sub_category** – Product sub-category
* **brand** – Product brand
* **discounted_price** – Current selling price of the product
* **original_price** – Original or listed price of the product
* **discount_percent** – Discount percentage
* **rating_out_of_5** – Product rating
* **rating_count** – Number of ratings or reviews associated with the product
* **platform** – Source platform of the product listing
* **shades** – Number of available shades, where applicable

Each row in the raw dataset represents a product listing collected from Nykaa. Since the same product may appear in multiple listings or categories, the cleaning process includes identifier extraction, duplicate investigation, and product-level deduplication.

During the cleaning process, additional columns such as `clean_url`, `product_id`, `sku_id`, and `product_key` are created to identify and standardize product records.

The final cleaned dataset is prepared for subsequent exploratory analysis and business analysis.


## Objectives

The objectives of this project are to:

* Clean and validate the raw product data scraped from Nykaa.
* Assess and handle missing values based on their relevance and meaning.
* Investigate and remove duplicate records and duplicate product listings.
* Create a product-level dataset using available product identifiers and URL-based information.
* Validate key data fields, including prices, discounts, ratings, and essential product information.
* Standardize data types and prepare the cleaned dataset for further analysis.
* Prepare the final dataset for subsequent exploratory analysis and business analysis using SQL, Excel, and Power BI.


## Contents

1. Import Libraries
2. Load Dataset
3. Initial Data Exploration
4. Data Cleaning

   * Missing Value Analysis and Treatment
   * Duplicate Record Investigation
   * Product Identifier Extraction
   * Product-Level Deduplication
   * Data Type Standardization
   * Price and Numeric Field Preparation
     
5. Logical Validation
6. Data Cleaning Summary
7. Export Cleaned Dataset
8. Conclusion


## 1. Import Libraries

The Pandas library is imported for data loading, cleaning, transformation, and analysis.


In [1]:
import pandas as pd

## 2. Load Dataset

The raw Nykaa product listing dataset is loaded from a CSV file into a Pandas DataFrame for initial exploration and subsequent data cleaning.


In [2]:
nykaa_df = pd.read_csv("nykaa_listing.csv")

The following initial exploration checks are performed to understand the structure, dimensions, data types, completeness, and numerical characteristics of the scraped dataset before the data-cleaning process begins.


## 3. Initial Data Exploration


### 3.1. Data Structure

The first few rows of the dataset are displayed to preview the data and verify its overall structure, including the available columns and the format of the product records.


In [3]:
nykaa_df.head()

,product_name,url,category,sub_category,brand,discounted_price,original_price,discount_percent,rating_out_of_5,rating_count,platform,shades
0,Aroma Magic Hand & Foot Care Combo,https://www.nykaa.com/aroma-magic-day-night-cr...,Hands Feet Care,Combo,Aroma Magic,315,450.0,30.0,4.0,7630.0,Nykaa,NaN
1,Bodyherbals Mani-Pedi Foot & Hand Care Combo,https://www.nykaa.com/bodyherbals-mani-pedi-fo...,Hands Feet Care,Combo,BodyHerbals,622,699.0,11.0,5.0,7.0,Nykaa,NaN
2,TNW The Natural Wash Hand & Foot Cream + Beetr...,https://www.nykaa.com/tnw-the-natural-wash-han...,Hands Feet Care,Combo,TNW The Natural Wash,389,648.0,40.0,4.0,6764.0,Nykaa,NaN
3,TNW The Natural Wash Under Arm Roll On and Han...,https://www.nykaa.com/tnw-the-natural-wash-und...,Hands Feet Care,Combo,TNW The Natural Wash,389,648.0,40.0,4.0,5166.0,Nykaa,NaN
4,TNW The Natural Wash Hand & Foot Cream + SPF 5...,https://www.nykaa.com/tnw-the-natural-wash-han...,Hands Feet Care,Combo,TNW The Natural Wash,539,898.0,40.0,4.0,4678.0,Nykaa,NaN


### 3.2. Checking Dataset Dimensions

The dataset dimensions are checked to determine the number of records and features in the raw dataset before data cleaning begins.


In [4]:
nykaa_df.shape

(27364, 12)

#### Observation

The raw dataset contains **27,364 records and 12 columns**, providing the baseline dataset size before the cleaning process.


### 3.3. Checking Data Types

The data type and non-null count of each column are examined to understand the structure of the dataset and identify columns that require further cleaning or conversion.


In [5]:
nykaa_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 27364 entries, 0 to 27363
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   product_name      27364 non-null  str    
 1   url               27364 non-null  str    
 2   category          27364 non-null  str    
 3   sub_category      27364 non-null  str    
 4   brand             27364 non-null  str    
 5   discounted_price  27364 non-null  int64  
 6   original_price    20223 non-null  float64
 7   discount_percent  20172 non-null  float64
 8   rating_out_of_5   25097 non-null  float64
 9   rating_count      25097 non-null  float64
 10  platform          27364 non-null  str    
 11  shades            4888 non-null   str    
dtypes: float64(4), int64(1), str(7)
memory usage: 2.5 MB


#### Observation

The dataset contains a combination of text, integer, and floating-point columns. The output also confirms that several columns, including `original_price`, `discount_percent`, `rating_out_of_5`, `rating_count`, and `shades`, contain missing values and require further investigation.


### 3.4. Summary Statistics

Summary statistics are used to understand the main numerical columns in the dataset. They help us check the average, minimum, maximum, and overall range of the values before starting the data-cleaning process. This can help identify unusual values that may need further checking.



In [6]:
nykaa_df.describe()

,discounted_price,original_price,discount_percent,rating_out_of_5,rating_count
count,27364.000000,20223.000000,20172.000000,25097.000000,2.509700e+04
mean,1138.030953,902.255748,18.921227,4.246245,3.800398e+04
std,2737.524577,1141.055498,11.695227,0.576137,3.972936e+05
min,10.000000,39.000000,1.000000,1.000000,1.000000e+00
25%,305.000000,349.000000,10.000000,4.000000,1.500000e+01
50%,549.000000,595.000000,15.000000,4.000000,1.190000e+02
75%,1200.000000,1005.000000,25.000000,5.000000,1.333000e+03
max,174500.000000,55500.000000,83.000000,5.000000,1.116496e+07


## 4. Data Cleaning

### 4.1. Missing Value Analysis and Treatment


In [7]:
nykaa_df.isnull().sum()

product_name            0
url                     0
category                0
sub_category            0
brand                   0
discounted_price        0
original_price       7141
discount_percent     7192
rating_out_of_5      2267
rating_count         2267
platform                0
shades              22476
dtype: int64

#### Observation

The dataset contains missing values in five columns:

* `original_price` – 7,141 missing values
* `discount_percent` – 7,192 missing values
* `rating_out_of_5` – 2,267 missing values
* `rating_count` – 2,267 missing values
* `shades` – 22,476 missing values

Each of these missing values will be checked and handled separately based on the type of information.


#### 4.1.1. Handling Missing Price Values

The first step is to handle the missing values in the price-related columns.

At first we can deal with missing prices.

###  Missing discount %
Next, we check the entries where both original_price and discounted_price are available, but discount_percent is missing.

For these entries, the missing discount percentage can be calculated using:

$$ \text{discount\_percent} = \frac{\text{original\_price} - \text{discounted\_price}}{\text{original\_price}} \times 100 $$

Or equivalently:

$$ \text{discount\_percent} = 100 - \left( \frac{\text{discounted\_price}}{\text{original\_price}} \times 100 \right) $$

In [8]:
nykaa_df[
    nykaa_df["original_price"].notna() & 
    nykaa_df["discounted_price"].notna() &  
    nykaa_df["discount_percent"].isna()]["product_name"].count()

np.int64(51)

#### Observation

There are **51 entries** where `original_price` and `discounted_price` are available, but `discount_percent` is missing.


#### Decision

For these 51 entries, the missing `discount_percent` is calculated using the available `original_price` and `discounted_price` values.


In [9]:
null_disc_percent = nykaa_df["original_price"].notna() & nykaa_df["discounted_price"].notna() &  nykaa_df["discount_percent"].isna()
#null_disc_percent

In [10]:
nykaa_df.loc[null_disc_percent, "discount_percent"] = 100 - (
    nykaa_df.loc[null_disc_percent, "discounted_price"] / 
    nykaa_df.loc[null_disc_percent, "original_price"]
) * 100


### 4.1.2. Handling Missing `original_price`

Next, we check the entries where `original_price` is missing. We compare these entries with `discounted_price` and `discount_percent` to understand why the values are missing and decide how they should be handled.


In [11]:
nykaa_df[
    nykaa_df["original_price"].isna()
][["discounted_price", "discount_percent"]].count()

discounted_price    7141
discount_percent       0
dtype: int64

#### Observation

There are **7,141 entries** with a missing `original_price`. All of these entries have a `discounted_price`, while none of them have a `discount_percent`.


The dataset is checked again to confirm whether any entry has a missing `original_price` but still contains a `discount_percent`.


In [12]:
nykaa_df[
    nykaa_df["original_price"].isna() &
    nykaa_df["discount_percent"].notna()
]

,product_name,url,category,sub_category,brand,discounted_price,original_price,discount_percent,rating_out_of_5,rating_count,platform,shades


#### Observation

There are no entries with a missing `original_price` and an available `discount_percent`. Therefore, all **7,141 entries** are missing both `original_price` and `discount_percent`.


#### Decision

Since `discounted_price` is available for these entries, we set:

* `original_price` equal to `discounted_price`
* `discount_percent` equal to `0`

This means the product is treated as having no discount when the original price is not available.


In [13]:
null_original_price = nykaa_df[
    "original_price"].isna() & nykaa_df[
        "discounted_price"].notna() &  nykaa_df[
        "discount_percent"].isna()

In [14]:
nykaa_df.loc[null_original_price , "original_price"] = nykaa_df.loc[null_original_price , "discounted_price"]
nykaa_df.loc[null_original_price , "discount_percent"] = 0

### 4.1.3. Checking Missing Ratings and `rating_count`

Next, we check the relationship between `rating_out_of_5` and `rating_count`. The purpose is to understand whether only one of these fields is missing or whether both are missing for the same products.


#### Missing only `rating_count`

We check whether any entry has a `rating_out_of_5` value but does not have a `rating_count`.


In [15]:
nykaa_df[nykaa_df["rating_out_of_5"].notna() &
    nykaa_df["rating_count"].isna()]["product_name"].count()

np.int64(0)

#### Observation

There are no entries where `rating_out_of_5` is available but `rating_count` is missing.


#### Missing only `rating_out_of_5`

We check whether any entry has a `rating_count` value but does not have a `rating_out_of_5` value.


In [16]:
nykaa_df[nykaa_df["rating_out_of_5"].isna() & nykaa_df["rating_count"].notna()]["product_name"].count()

np.int64(0)

#### Observation

There are no entries where `rating_count` is available but `rating_out_of_5` is missing.


#### Missing both `rating_out_of_5` and `rating_count`

We now check how many entries are missing both rating fields.


In [17]:
nykaa_df[nykaa_df["rating_out_of_5"].isna() &
    nykaa_df["rating_count"].isna()]["product_name"].count()

np.int64(2267)

In [18]:
nykaa_df[nykaa_df["rating_out_of_5"].notna() &
    nykaa_df["rating_count"].notna()]["product_name"].count()

np.int64(25097)

#### Observation

There are **2,267 entries** where both `rating_out_of_5` and `rating_count` are missing. The remaining **25,097 entries** contain both values.

This shows that the two rating fields are missing together rather than separately.


A sample of products with missing rating information is displayed to understand the type of records where these values are not available.


In [19]:
nykaa_df[
    nykaa_df["rating_out_of_5"].isna() &
    nykaa_df["rating_count"].isna()
].sample(5)

,product_name,url,category,sub_category,brand,discounted_price,original_price,discount_percent,rating_out_of_5,rating_count,platform,shades
5675,Estee Lauder Re-Nutriv Ultimate Lift Regenerat...,https://www.nykaa.com/estee-lauder-re-nutriv-u...,Skin Care,Face Moisturizer,Estee Lauder,14400,14400.0,0.0,NaN,NaN,Nykaa,NaN
13417,inde wild Sun Defense Mineral Sunscreen SPF 50...,https://www.nykaa.com/inde-wild-sun-defense-mi...,Skin Care,Face Sunscreen,inde wild,959,1199.0,20.0,NaN,NaN,Nykaa,NaN
15898,it's a 10 Haircare Miracle Daily Conditioner -...,https://www.nykaa.com/it-s-a-10-haircare-mirac...,Hair Care,Conditioner,it's a 10 Haircare,1512,1890.0,20.0,NaN,NaN,Nykaa,NaN
27160,SOEZI Heeramandi Toezi Reusable Press On Nails...,https://www.nykaa.com/soezi-heeramandi-toezi-r...,Nail Makeup,Nail Art Kits,SOEZI,1699,1699.0,0.0,NaN,NaN,Nykaa,NaN
14451,Kerastase Genesis Homme Thickening Shampoo For...,https://www.nykaa.com/kerastase-genesis-homme-...,Hair Care,Shampoo,Kerastase,2900,2900.0,0.0,NaN,NaN,Nykaa,NaN


#### Decision

The **2,267 products** without rating information are retained with missing values. Since both `rating_out_of_5` and `rating_count` are missing together, there is no reliable value available in the dataset to fill them.

Therefore, these missing values are left as they are.


### 4.2. Deal with Duplicates

Duplicate records can occur in scraped data when the same product is collected more than once. In this section, we check for duplicate rows, duplicate URLs, and duplicate product names.

We then use product_id, sku_id, and product_key to understand whether repeated records are true duplicates or different product variants.


#### 4.2.1. Checking Duplicate Records

First, we check for completely identical rows in the dataset. This helps us identify records where all column values are exactly the same.



In [20]:
pd.set_option('display.max_colwidth', None)

In [21]:
nykaa_df[nykaa_df.duplicated()]

,product_name,url,category,sub_category,brand,discounted_price,original_price,discount_percent,rating_out_of_5,rating_count,platform,shades


Observation

There are no completely duplicate rows in the dataset. This means that no two rows have exactly the same values across all columns.

#### 4.2.2. Checking Duplicate URLs

Even if complete duplicate rows are not present, the same product may still appear more than once. Therefore, we check the url column for duplicate URLs.


In [22]:
nykaa_df[nykaa_df['url'].duplicated()]

,product_name,url,category,sub_category,brand,discounted_price,original_price,discount_percent,rating_out_of_5,rating_count,platform,shades


In [23]:
nykaa_df["url"].nunique()

27364

#### Observation

There are no duplicate URLs in the dataset. The dataset contains 27,364 unique URLs for 27,364 rows.

This means that URL alone can distinguish the scraped listing records, but it does not necessarily mean that every URL represents a different product.

### 4.2.3. Checking Duplicate Product Names

Next, we check whether the same product name appears in multiple rows.

In [24]:
nykaa_df[nykaa_df['product_name'].duplicated()]

,product_name,url,category,sub_category,brand,discounted_price,original_price,discount_percent,rating_out_of_5,rating_count,platform,shades
5,Aroma Magic Hand & Foot Care Combo,https://www.nykaa.com/aroma-magic-day-night-cream-combo/p/539653?productId=539653&pps=13,Hands Feet Care,Combo,Aroma Magic,338,450.0,25.0,4.0,7632.0,Nykaa,NaN
8,Moha Hand And Foot Care Combo Heals Cracked Skin And Nourishes For Smooth Healthy Hands And Feet,https://www.nykaa.com/moha-hand-and-feet-care-combo/p/664854?productId=664854&pps=6,Hands Feet Care,Combo,Moha,356,475.0,25.0,4.0,3599.0,Nykaa,NaN
12,Bodyherbals Mani-Pedi Foot & Hand Care Combo,https://www.nykaa.com/bodyherbals-mani-pedi-foot-hand-care-combo-tea-tree-oil-foot-scrub-grape-fruit-seed-hand-scrub/p/6008580?productId=6008580&pps=3,Hands Feet Care,Combo,BodyHerbals,664,699.0,5.0,5.0,7.0,Nykaa,NaN
14,TNW The Natural Wash Hand & Foot Cream + SPF 50 Sunscreen Combo,https://www.nykaa.com/tnw-the-natural-wash-hand-foot-cream-spf-50-sunscreen-combo/p/3877653?productId=3877653&pps=10,Hands Feet Care,Combo,TNW The Natural Wash,584,898.0,35.0,4.0,4680.0,Nykaa,NaN
53,L'Occitane Almond Hand Cream,https://www.nykaa.com/l-occitane-almond-delicious-hands/p/1375675?productId=1375675&skuId=286267&pps=18,Hands Feet Care,Hand Creams,L'Occitane Skincare,1050,1050.0,0.0,5.0,264.0,Nykaa,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
27338,"Bella Cotton Pads With D Panthenol & Orange Blossom Extracts, 100% Cotton, Double Sided",https://www.nykaa.com/bella-cotton-pads-round-with-d-panthenol-orange-blossom-extract-70-pieces/p/175546?productId=175546&pps=9,Nail Makeup,Nail Polish Remover,Bella,114,130.0,12.0,4.0,1132.0,Nykaa,NaN
27342,"Bella Cotton Balls, 100% Cotton, Soft, Highly Absorbent,Makeup Removal, Nail Polish Removal",https://www.nykaa.com/bella-cosmetic-cotton-balls-coloured-a100/p/23764?productId=23764&pps=13,Nail Makeup,Nail Polish Remover,Bella,150,168.0,11.0,5.0,199.0,Nykaa,NaN
27343,Fabskin Cosmetic Cotton Balls - 100,https://www.nykaa.com/fabskin-cosmetic-cotton-balls-100/p/7997556?productId=7997556&pps=14,Nail Makeup,Nail Polish Remover,Fabskin,129,140.0,8.0,4.0,9.0,Nykaa,NaN
27353,"Bella Cotton Pads With D Panthenol&Orange Blossom Extracts,100% Cotton,Double Sided - Pack of 2",https://www.nykaa.com/bella-cotton-pads-with-d-panthenol-orange-pack-of-2/p/1505369?productId=1505369&pps=4,Nail Makeup,Nail Polish Remover,Bella,234,260.0,10.0,4.0,2264.0,Nykaa,NaN


In [25]:
nykaa_df[nykaa_df['product_name'].duplicated()]["product_name"].count()

np.int64(4423)

#### Observation

There are 4,423 rows with duplicate product names.

However, duplicate product names do not always mean duplicate products. The same product can appear under different categories or can have different variants. Therefore, product name alone should not be used to remove duplicates.

To investigate this further, we extract the product ID from the URL.

#### Extract Product ID

Since some product names appear more than once, the next step is to extract the product_id from the product URL.

The product_id helps us identify the main product associated with each listing and check whether the same product appears multiple times in the dataset.

In [26]:
nykaa_df['product_id'] = nykaa_df['url'].str.extract(r'productId=(\d+)')

### Identifying Products with Multiple Listings

After extracting product_id, we count how many times each product ID appears.

This helps us identify products that have been collected multiple times from different category pages or different URL paths.

In [27]:
duplicate_product_ids = (
    nykaa_df['product_id']
    .value_counts()
)

Find only product_ids appearing more than once

In [28]:
duplicate_product_ids = duplicate_product_ids[duplicate_product_ids > 1]

duplicate_product_ids

product_id
25011560    24
418274      22
13192269    22
21112351    16
347977      15
            ..
860868       2
484941       2
25408296     2
3445863      2
26615027     2
Name: count, Length: 3441, dtype: int64

total number of duplicate listings

In [29]:
print(f"Unique product IDs with multiple listings: {len(duplicate_product_ids):,}")
print(f"Total occurrences of these product IDs: {duplicate_product_ids.sum():,}")
print(f"Duplicate listings (excluding first occurrence): {(duplicate_product_ids - 1).sum():,}")

Unique product IDs with multiple listings: 3,441
Total occurrences of these product IDs: 8,024
Duplicate listings (excluding first occurrence): 4,583


#### Observation

There are 3,441 product IDs that appear more than once.

These product IDs account for 8,024 total listings, resulting in 4,583 duplicate listings after excluding the first occurrence of each product ID.

This shows that the same product can appear in multiple listings in the scraped dataset.

### Investigation of Product IDs and SKUs

A product ID can be linked to more than one sku_id. This indicates that one product can have different variants, such as different shades, sizes, or pack options.

Therefore:

product_id represents the main product or product family.
sku_id represents a specific product variant when it is available.

Because of this, using only product_id for deduplication could remove valid product variants.

### Extracting SKU IDs

To understand the repeated product listings more clearly, we extract the sku_id from the URL.

The sku_id is then compared with the product_id to understand whether repeated listings are different variants of the same product.

In [30]:
nykaa_df['sku_id'] = nykaa_df['url'].str.extract(r'skuId=(\d+)')

In [31]:
nykaa_df.groupby('product_id')['sku_id'].nunique().sort_values(ascending=False)

product_id
25011560    23
418274      21
347977      13
21112351    10
23053290    10
            ..
1003709      0
1003710      0
1003711      0
1003712      0
10017257     0
Name: sku_id, Length: 22781, dtype: int64

#### Observation: Product vs. SKU Hierarchy

Analysis reveals that some duplicate entries sharing the same `product_id` have different `sku_id` values. This indicates that a single product can have multiple variants, such as different shades, sizes, or pack options.

Therefore:

- `product_id` represents the main product or product family.
- `sku_id` represents a specific product variant when available.

#### Verifying SKU ID Uniqueness Across Products

We also check whether the same sku_id is linked to more than one product_id.

In [32]:
sku_product_map = (
    nykaa_df.groupby('sku_id')['product_id']
    .nunique()
)

sku_product_map[sku_product_map > 1]

Series([], Name: product_id, dtype: int64)

#### Observation

No sku_id was found under more than one product_id.

This supports the use of sku_id as a more detailed identifier when it is available.

#### Investigation of Missing SKU IDs

Most records in the dataset do not contain a sku_id.

#### Observation

Out of 27,364 rows, 26,679 rows do not have a sku_id.

Because of this, product_id is used as a fallback when creating the product_key. This can cause the same product to appear multiple times when it is listed under different categories.

The sample below shows one such case, where the same product has the same price and rating but appears under different categories and different pps values in the URL.

### Decision: Creating a Unified Product Key

Since sku_id is not available for most records, we create a unified product_key.

The logic is:

Use sku_id when it is available.
Otherwise, use product_id.

This allows us to keep variant-level information where a SKU is available while still identifying products that do not have a SKU.

#### Creating a Unified Product Identifier

To unify the available product identifiers, a `product_key` column is created. The logic uses `sku_id` when it is available; otherwise, it falls back to `product_id`.

This allows SKU-level identification where available while providing a fallback identifier for records without a SKU.

In [33]:
nykaa_df['product_key'] = (
    nykaa_df['sku_id']
    .fillna(nykaa_df['product_id'])
)

### Checking Duplicate Product Keys

After creating product_key, we check how many product keys still appear more than once.

In [34]:
nykaa_df[nykaa_df['product_key'].duplicated()]["product_name"].count()

np.int64(3989)

In [35]:
duplicate_ids = (
    nykaa_df['product_key']
    .value_counts()
)
duplicate_product_key = duplicate_ids[duplicate_ids > 1]

duplicate_product_key

product_key
10980626    9
10678619    8
12255320    8
10846914    7
1153924     6
           ..
860868      2
484941      2
25408296    2
3445863     2
26615027    2
Name: count, Length: 3215, dtype: int64

In [36]:
print(f"product_key with multiple listings: {len(duplicate_product_key):,}")
print(f"Total occurrences of these product_key: {duplicate_product_key.sum():,}")
print(f"Duplicate listings (excluding first occurrence): {(duplicate_product_key - 1).sum():,}")

product_key with multiple listings: 3,215
Total occurrences of these product_key: 7,204
Duplicate listings (excluding first occurrence): 3,989


#### Observation

There are 3,215 product keys with multiple listings.

These account for 7,204 total occurrences, with 3,989 duplicate listings after excluding the first occurrence of each product key.

Further investigation shows that many of these repeated records are the same product listed under different categories or different URL tracking parameters.

#### Investigation: Identifying the Cause of Remaining Duplicates

Despite creating the unified `product_key`, some duplicate product keys remain. This occurs mainly because most records do not contain a `sku_id` and therefore use `product_id` as the fallback identifier.

The remaining duplicate records are investigated using product details and URL parameters to determine whether they represent different variants or repeated listings.

In [37]:
null_sku = nykaa_df[nykaa_df['sku_id'].isna()]

In [38]:
null_sku['product_name'].count()

np.int64(26679)

#### Observation: Missing SKU IDs

Out of **27,364 total entries, 26,679 rows do not contain a `sku_id`**.

Because `sku_id` is unavailable for most records, `product_id` is used as the fallback when creating `product_key`.

A sample of these records is examined to determine whether the repeated product IDs represent different variants or repeated listings across categories.

In [39]:
sample_pid = '10980626'  # or another duplicated product_id without sku_id

nykaa_df[nykaa_df['product_id'] == sample_pid][[
    'product_name',
    'category',
    'sub_category',
    'discounted_price',
    'original_price',
    'rating_out_of_5',
    'rating_count',
    'url'
]]

,product_name,category,sub_category,discounted_price,original_price,rating_out_of_5,rating_count,url
189,Natural Vibes Aloe Vera Gel,Hands Feet Care,Foot Creams,449,449.0,4.0,102.0,https://www.nykaa.com/natural-vibes-aloe-vera-gel/p/10980626?productId=10980626&pps=19
1234,Natural Vibes Aloe Vera Gel,Body Care,Body Sunscreen,449,449.0,4.0,102.0,https://www.nykaa.com/natural-vibes-aloe-vera-gel/p/10980626?productId=10980626&pps=1
5658,Natural Vibes Aloe Vera Gel,Skin Care,Face Moisturizer,449,449.0,4.0,102.0,https://www.nykaa.com/natural-vibes-aloe-vera-gel/p/10980626?productId=10980626&pps=20
6305,Natural Vibes Aloe Vera Gel,Skin Care,Night Cream,449,449.0,4.0,102.0,https://www.nykaa.com/natural-vibes-aloe-vera-gel/p/10980626?productId=10980626&pps=15
6747,Natural Vibes Aloe Vera Gel,Skin Care,All Purpose Gels/Creams,449,449.0,4.0,102.0,https://www.nykaa.com/natural-vibes-aloe-vera-gel/p/10980626?productId=10980626&pps=10
7986,Natural Vibes Aloe Vera Gel,Skin Care,Serums & Essence,449,449.0,4.0,102.0,https://www.nykaa.com/natural-vibes-aloe-vera-gel/p/10980626?productId=10980626&pps=2
13710,Natural Vibes Aloe Vera Gel,Skin Care,Face Sunscreen,449,449.0,4.0,102.0,https://www.nykaa.com/natural-vibes-aloe-vera-gel/p/10980626?productId=10980626&pps=6
18480,Natural Vibes Aloe Vera Gel,Eye Care,Under Eye Creams & Serums,449,449.0,4.0,102.0,https://www.nykaa.com/natural-vibes-aloe-vera-gel/p/10980626?productId=10980626&pps=14
20491,Natural Vibes Aloe Vera Gel,Face Makeup,Skin Tint,449,449.0,4.0,102.0,https://www.nykaa.com/natural-vibes-aloe-vera-gel/p/10980626?productId=10980626&pps=7


### Observation: Root Cause of the Remaining Duplicates

The sample investigation shows that some repeated product records have:

- The same product_id
- The same product name
- The same price
- The same rating information
- The same brand
- Different categories or sub-categories
- Different pps values in the URL

This indicates that these rows are repeated listings of the same product rather than separate products.

### Decision: Final Product-Level Deduplication

The main purpose of this project is to create a product-level dataset.

Therefore, repeated records with the same product_key are removed, while the first occurrence is retained.

The cleaned product-level dataset is stored separately as df_products.

The original nykaa_df is kept unchanged as the raw working dataset, so the original scraped records are still available if listing-level analysis is needed later.

In [40]:
df_products = nykaa_df.drop_duplicates(subset='product_key')

## 5. Logical Validation

After cleaning the missing values and duplicates, the dataset is checked using logical rules.

These checks help confirm that important values such as prices, discounts, ratings, and product identifiers are reasonable and consistent.

### 5.1. Price Validation

The price columns are checked to make sure that the values follow a logical relationship.

For a discounted product:

discounted_price should not be greater than original_price.
Both price values should be greater than zero.

These checks help identify incorrect price values before further analysis.

In [41]:
(df_products['discounted_price'] <= 0).sum()

np.int64(0)

In [42]:
(df_products['original_price'] <= 0).sum()

np.int64(0)

#### Observation

The price validation checks do not show any invalid price records.

This means the price values follow the expected relationship in the cleaned dataset.

### 5.2. Discount Validation

The discount percentage is checked to make sure that it is within a valid range.

A discount percentage should be between 0% and 100%.

Values below 0% or above 100% would indicate an invalid discount value.

In [43]:
df_products[df_products["discounted_price"] > df_products["original_price"]]

,product_name,url,category,sub_category,brand,discounted_price,original_price,discount_percent,rating_out_of_5,rating_count,platform,shades,product_id,sku_id,product_key


#### Observation

The validation check does not find any discount percentages outside the valid range.

Therefore, the discount values are within the expected range.

### 5.3. Discount Percentage Validation

The `discount_percent` column is checked to ensure that all values fall within the valid range of **0% to 100%**.

Values below 0% or above 100% would indicate invalid discount data.

In [44]:
(df_products['discount_percent'] < 0).sum()

np.int64(0)

In [45]:
(df_products['discount_percent'] > 100).sum()

np.int64(0)

#### Observation

All `discount_percent` values fall within the valid range of 0% to 100%.

No negative discount values or values exceeding 100% were found.

### 5.4. Discount Percentage Cross-Validation

This step performs a crucial cross-validation check to ensure internal data consistency. We are verifying that the **scraped discount percentage** provided by the platform matches our own **calculated discount percentage** derived mathematically from the `original_price` and `discounted_price`. 

Because rounding variations are common on e-commerce platforms (e.g., a calculated 14.7% being displayed as 15%), we will allow a small, acceptable tolerance margin rather than demanding a perfect match. Large discrepancies, however, would flag potential data corruption or a fundamental misunderstanding of how the platform computes markdown values.

In [46]:
calculated_discount = (
    (df_products['original_price'] - df_products['discounted_price'])
    / df_products['original_price'] * 100
).round()

(df_products['discount_percent'] != calculated_discount).sum()

np.int64(156)

In [47]:
calculated_discount = (
    ((df_products['original_price'] - df_products['discounted_price'])
     / df_products['original_price'] * 100)
).round()

discount_check = df_products.copy()
discount_check['calculated_discount'] = calculated_discount

mismatches = discount_check[
    discount_check['discount_percent'] != discount_check['calculated_discount']
]

mismatches[[
    'product_name',
    'original_price',
    'discounted_price',
    'discount_percent',
    'calculated_discount'
]].head(10)

,product_name,original_price,discounted_price,discount_percent,calculated_discount
55,TNW The Natural Wash Coconut Hand & Foot Cream For Super Soft and Moisturized Skin - Non- sticky,349.0,239,31.000000,32.0
154,Himalaya Wellness Foot Care Cream,140.0,131,6.428571,6.0
423,Faces Canada Lip Balm 12hr Moisture For Dry - Chapped Lips,175.0,139,20.000000,21.0
926,Sanfe Hydrogel Lip Mask - Pack Of 1,49.0,47,5.000000,4.0
969,Johnson's Baby Milk + Rice Combo,465.0,449,4.000000,3.0
1087,SkinInspired Invishield 365 Body Sunscreen Spray SPF 80+ PA++++,799.0,779,2.000000,3.0
1824,Keya Seth Aromatherapy Medicure Ankush Antiseptic Liquid,65.0,59,10.000000,9.0
2523,Aroma Magic Anti Bact Soothing Cream,95.0,86,10.000000,9.0
2755,Vivel Lavender & Almond Oil Body Wash,45.0,40,11.111111,11.0
2830,Yardley London Morning Dew Shower Gel,259.0,225,13.127413,13.0


In [48]:
mismatches['difference'] = (
    mismatches['discount_percent'] -
    mismatches['calculated_discount']
)

mismatches['difference'].value_counts().sort_index()

difference
-45.000000     1
-27.000000     2
-25.000000     1
-24.000000     1
-21.000000     1
-19.000000     1
-13.000000     3
-12.000000     1
-11.000000     1
-10.000000     1
-6.000000      1
-5.000000      2
-4.000000      1
-3.000000      2
-2.000000      1
-1.000000     25
-0.492183      2
-0.484355      1
-0.443760      1
-0.433962      1
-0.400000      1
-0.333333      2
-0.333333      3
-0.285714      1
-0.267559      1
-0.223022      1
-0.210084      1
-0.206897      1
-0.125628      1
-0.081081      1
-0.027027      1
 0.005003      1
 0.015019      1
 0.020050      1
 0.033445      1
 0.050167      1
 0.090909      1
 0.093633      2
 0.111111      1
 0.127413      1
 0.152866      1
 0.333333      2
 0.388982      1
 0.400000      1
 0.408163      1
 0.428571      1
 0.474576      1
 0.484355      1
 1.000000     73
Name: count, dtype: int64

In [49]:
discount_check['difference'] = (
    discount_check['discount_percent'] -
    discount_check['calculated_discount']
)

major_mismatch = discount_check[
    discount_check['difference'].abs() > 1
]

len(major_mismatch)

20

#### Observation: Uncovering Discount Discrepancies

To ensure our metrics tell a consistent story, I cross-validated the scraped discount percentages against a fresh mathematical calculation derived straight from the `original_price` and `discounted_price` columns. 

Here is what the data revealed:
* **Minor Rounding Differences:** The vast majority of slight variances (**136 records**) fell within a tiny ±1 percentage point window. These are completely normal and can be safely attributed to how the platform rounds display percentages on its front-end.
* **True Inconsistencies:** However, **20 products** showed significant discrepancies where the displayed discount completely drifted from the actual price values.

**The Fix:** Because the discount percentage is inherently a derived metric, I chose to recalculate it across the board using the raw price columns. This ensures absolute mathematical consistency across the entire dataset before we begin modeling or visualization.

### Decision: Preserving Both Discount Metrics

Instead of overwriting the original data, I decided to keep the raw, scraped discount percentage as-is and create a brand new column named `calculated_discount_percentage`. 

**Why this approach?**
* **Preserving Front-End Reality:** The scraped value reflects exactly what the customer saw on the website, which is invaluable for any future behavioral or conversion rate analysis.
* **Ensuring Mathematical Accuracy:** The new calculated column provides an uncompromised, exact mathematical metric derived strictly from our price points, giving us the clean data we need for modeling and financial analysis.

### 5.5. Engineering the Calculated Discount Percentage

The original scraped discount percentage is preserved in `discount_percent_scraped`.

The `discount_percent` column is then recalculated from the original and discounted prices using the standard discount formula:

$$
Discount\ \% =
\frac{Original\ Price - Discounted\ Price}
{Original\ Price}
\times 100
$$

This provides a mathematically consistent discount percentage for downstream analysis while preserving the original scraped value for reference.

In [50]:
df_products['discount_percent_scraped'] = df_products['discount_percent']

df_products['discount_percent'] = (
    ((df_products['original_price'] - df_products['discounted_price'])
     / df_products['original_price'] * 100)
).round().astype(float)

### 5.6. Rating Validation

The `rating_out_of_5` column is checked to ensure that ratings fall within the valid range of **0 to 5 stars**.

Values outside this range would indicate invalid or incorrectly parsed rating data.

In [51]:
df_products[
    (df_products['rating_out_of_5'] < 0) |
    (df_products['rating_out_of_5'] > 5)
]

,product_name,url,category,sub_category,brand,discounted_price,original_price,discount_percent,rating_out_of_5,rating_count,platform,shades,product_id,sku_id,product_key,discount_percent_scraped


#### Observation

No product ratings were found outside the valid range of 0 to 5 stars.

### 5.7. Rating Count Validation

The `rating_count` column is checked to ensure that review counts are not negative.

Negative review counts would indicate invalid or incorrectly parsed data.

In [52]:
(df_products['rating_count'] < 0).sum()

np.int64(0)

#### Observation

No entries have a negative `rating_count`, indicating that the review-count values satisfy the basic non-negative constraint.

### 5.8. Duplicate Product Key Validation

This final logical check ensures the integrity of our dataset's grain. We are verifying that our primary identifier — the **product key** (or unique identifier) — contains absolutely zero duplicate entries. 

Duplicate keys at this stage would mean that our previous deduplication phase missed edge cases, or that the engineered key isn't unique enough to separate distinct items. Guaranteeing a strict 1:1 relationship between each row and a unique product key is essential before we perform any aggregations or group-by operations in our exploratory analysis.

In [53]:
df_products['product_key'].duplicated().sum()

np.int64(0)

#### Observation: Product Key Uniqueness Confirmed

The integrity check passed perfectly: **the product key is entirely unique for every single product in the dataset.** 

There are zero duplicate records or overlapping identifiers, confirming that the fundamental grain of our data is clean. With a verified 1:1 mapping for each row, we can safely proceed to aggregations, merges, and downstream analysis without the risk of inflating our row counts.

### 5.9. Missing Essential Fields Validation

The dataset is checked for missing values in the essential fields required for product-level analysis:

- `product_name`
- `brand`
- `category`
- `discounted_price`

These fields are required for identifying and analyzing products in the downstream analysis.

In [54]:
essential_cols = [
    'product_name',
    'brand',
    'category',
    'discounted_price'
]

df_products[essential_cols].isna().sum()

product_name        0
brand               0
category            0
discounted_price    0
dtype: int64

#### Observation

No missing values were found in the designated essential fields. Therefore, all records contain the minimum information required for downstream product-level analysis.

### 5.10. Product Category Sanity Check

This step verifies the logical hierarchy of our product taxonomy. By auditing **how many unique subcategories exist under each main category**, we can ensure that the mapping makes structural sense. 

This sanity check helps us spot data entry errors, such as a subcategory accidentally nested under the wrong main branch, spelling inconsistencies that split a single category into two, or unusually fragmented taxonomies that might complicate our high-level aggregations and reporting.

In [55]:
df_products.groupby('category')['sub_category'].nunique().sort_values(ascending=False)

category
Face Makeup         17
Skin Care           16
Eye Makeup          10
Body Care            9
Hair Care            9
Lip Makeup           8
Hands Feet Care      4
Lip Care             4
Nail Makeup          4
Eye Care             3
Eye & Lip Makeup     1
Name: sub_category, dtype: int64

#### Observation: Essential Fields Completeness Confirmed

The validation check passed flawlessly: **there are absolutely no missing values within the designated essential fields.** 

Core columns such as `product_key`, `original_price`, `brand`, and `category` are 100% populated across the entire dataset. This complete data density ensures that our structural backbone is secure and eliminates the need for any complex null-imputation strategies before moving into exploration.

### 5.11. Final Validation

The cleaned dataset passed the structural and logical validation checks performed in this notebook.

The validation covered:

- Price values
- Discount percentages
- Discount cross-validation
- Product ratings
- Rating counts
- Product-key uniqueness
- Essential-field completeness
- Category/subcategory structure

The validated dataset is ready for downstream exploratory analysis and business analysis.

## 6. Convert Appropriate Data Types

The columns are converted to appropriate data types to support analysis, calculations, and downstream SQL operations.

### 6.1. Convert Shade Count to Numeric

The `shades` column contains text values such as `"25 Shades"`. Since the column represents the number of available shades, the numeric portion is extracted and converted to an integer-compatible data type.

### 6.1.1. Extract numeric shade count
The shades column contains text values such as "25 Shades". The numeric shade count is extracted to enable quantitative analysis.

In [56]:
cleaned_series = df_products['shades'].astype(str).str.replace(r'\D+', '', regex=True)
cleaned_series = cleaned_series.replace('', None)

### 6.1.2. Convert to numeric data type
After extraction, the column is converted to an integer data type so it can be used for statistical analysis and stored appropriately in SQL.

In [57]:
df_products['shades'] = pd.to_numeric(cleaned_series, errors='coerce').astype("Int64")

In [58]:
df_products['shades'].dtype

Int64Dtype()

In [59]:
df_products.rename(columns={'shades':'shade_count'}, inplace=True)

During validation of the shades column, a small number of records were found to contain product names instead of shade counts due to inconsistent page structures. These records were verified against the product pages and corrected manually.

## 7. Round Numeric Columns

Before exporting the cleaned dataset, selected numerical columns are rounded to two decimal places to maintain consistent numerical formatting.

In [60]:
df_products[['discounted_price', 
    'original_price',
    'discount_percent',
    'discount_percent_scraped']] = (
    df_products[['discounted_price', 
        'original_price',
        'discount_percent',
        'discount_percent_scraped']]
        .round(2)
)

## 8. Data Cleaning Summary
| Step | Result |
| --- | --- |
| Missing values handled | ✔ |
| Missing prices imputed | ✔ |
| Missing ratings retained | ✔ |
| Duplicate listings investigated | ✔ |
| Product key created | ✔ |
| Product-level duplicates removed | ✔ |
| Price validation completed | ✔ |
| Discount validation completed | ✔ |
| Rating validation completed | ✔ |
| Shade count converted to numeric | ✔ |
| Numeric columns rounded | ✔ |

Final cleaned dataset contains **23,375 products** ready for downstream analysis.

## 9. Export Cleaned Dataset

The final cleaned product-level dataset is exported to CSV for use in the subsequent exploratory analysis, SQL analysis, dashboard development, and cross-platform analysis.

In [61]:
print(df_products.shape)

(23375, 16)


In [62]:
print(df_products.head())

                                                         product_name  \
0                                  Aroma Magic Hand & Foot Care Combo   
1                        Bodyherbals Mani-Pedi Foot & Hand Care Combo   
2    TNW The Natural Wash Hand & Foot Cream + Beetroot Lip Balm Combo   
3  TNW The Natural Wash Under Arm Roll On and Hand & Foot Cream Combo   
4     TNW The Natural Wash Hand & Foot Cream + SPF 50 Sunscreen Combo   

                                                                                                                                                      url  \
0                                                                https://www.nykaa.com/aroma-magic-day-night-cream-combo/p/539653?productId=539653&pps=12   
1  https://www.nykaa.com/bodyherbals-mani-pedi-foot-hand-care-combo-tea-tree-oil-foot-scrub-grape-fruit-seed-hand-scrub/p/6008580?productId=6008580&pps=8   
2                                   https://www.nykaa.com/tnw-the-natural-wash-hand-foot-c

In [63]:
df_products.to_csv("nykaa_products.csv",index=False)

## 10. Conclusion

The scraped Nykaa dataset was cleaned and validated through missing value treatment, duplicate investigation, identifier extraction, and business rule validation. The resulting dataset is internally consistent and ready for exploratory analysis, SQL analysis, dashboard development, and future cross-platform comparison. This cleaned dataset will be used in subsequent notebooks for SQL queries, exploratory data analysis, dashboard creation, and future cross-platform comparisons.